crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!     .partitioned_rolling_aggregate_linear(
+//!         |vaxxed| *vaxxed,
+//!         |total| total,
+//!         RelRange::new(RelOffset::Before(u32::MAX), RelOffset::Before(0)),
+//!     )
+//! ```
+//!
+//! Then we join against the population table, which also requires a re-index
+//! step:
+//!
+//! ```ignore
+//!         .map_index(|(c, (date, total))| {
+//!             (
+//!                 c.clone(),
+//!                 ((date / 12) as i32, (date % 12 + 1) as u8, total.unwrap()),
+//!             )
+//!         })
+//!         .join_index(&pop_stream, |c, (y, m, total), pop| {
+//!             Some(((c.clone(), *y, *m), (*total, *pop)))
+//!         })
+//! ```
+//!
+//! And finally we can print the results:
+//!
+//! ```ignore
+//!         .inspect(|data| {
+//!             data.iter().for_each(|((c, y, m), (total, pop), w)| {
+//!                 let rate = total as f64 / pop as f64 * 100.0;
+//!                 println!("{c:16} {y}-{m:02}: {total:9} {pop:8} {rate:5.2}% {w:+}",)
+//!             })
+//!         });
+//! ```
+//!
+//! The complete program is in `tutorial9.rs`.  If we run it, it prints, in
+//! part, the following. The percentages over 100% are correct because this data
+//! counts vaccination doses rather than people:
+//!
+//! ```text
+//! England          2021-01:   5600174 56286961  9.95% +1
+//! England          2021-02:  14977592 56286961 26.61% +1
+//! England          2021-03:  26838767 56286961 47.68% +1
+//! England          2021-04:  38127712 56286961 67.74% +1
+//! England          2021-05:  51900658 56286961 92.21% +1
+//! England          2021-06:  62845573 56286961 111.65% +1
+//! ...
+//! ```
+//!
+//! ## Output
+//!
+//! We've done a lot of data processing in DBSP and we've printed out results,
+//! but all that printing happened within the circuit.  We didn't ship the
+//! results outside the circuit to the caller's context.  That's the final step.
+//!
+//! It's easy.  [`Stream::output`] returns an [`OutputHandle`] that one can
+//! return to the caller, which can then read out the data after each step.
+//!
+//! We can change `build_circuit()` to return its output instead of printing it.
+//! We add an [`OutputHandle`] to its return type:
+//!
+//! ```ignore
+//! fn build_circuit(
+//!     circuit: &mut RootCircuit,
+//! ) -> Result<(
+//!     CollectionHandle<Record, isize>,
+//!     CollectionHandle<String, (u64, isize)>,
+//!     OutputHandle<OrdIndexedZSet<(String, i32, u8), (isize, u64), isize>>
+//! )> {
+//! ...
+//! }
+//! ```
+//!
+//! and then return the handle instead of `inspect`ing it:
+//!
+//! ```ignore
+//!     let rates = monthly_totals
+//! ...
+//!         .join_index(&pop_stream, |c, (y, m, total), pop| {
+//!             Some(((c.clone(), *y, *m), (*total, *pop)))
+//!         });
+//!     Ok((vax_handle, pop_handle, rates.output()))
+//! ```
+//!
+//! Finally, in `main()`, we read the data out of the [`OutputHandle`] after
```

Comment:
This bit was undocumented in the SQL compiler README.
I need to create something similar there.